File: examples/core/sample_framebuffer.cpp

#include <CEngine.h>

#include <core/CFrameBuffer.h>

int main()
{
    auto _app = new engine::COpenGLApp();
    _app->init();

    engine::CAttachmentConfig _fbColorConfig;
    _fbColorConfig.name                 = "color_attachment";
    _fbColorConfig.attachment           = engine::eFboAttachment::COLOR;
    _fbColorConfig.width                = engine::COpenGLApp::GetWindow()->width();
    _fbColorConfig.height               = engine::COpenGLApp::GetWindow()->height();
    _fbColorConfig.texInternalFormat    = engine::eTextureFormat::RGB;
    _fbColorConfig.texFormat            = engine::eTextureFormat::RGB;
    _fbColorConfig.texPixelDataType     = engine::ePixelDataType::UINT_8;
    _fbColorConfig.texWrapU             = engine::eTextureWrap::REPEAT;
    _fbColorConfig.texWrapV             = engine::eTextureWrap::REPEAT;

    engine::CAttachmentConfig _fbDepthConfig;
    _fbDepthConfig.name                 = "depth_attachment";
    _fbDepthConfig.attachment           = engine::eFboAttachment::DEPTH;
    _fbDepthConfig.width                = engine::COpenGLApp::GetWindow()->width();
    _fbDepthConfig.height               = engine::COpenGLApp::GetWindow()->height();
    _fbDepthConfig.texInternalFormat    = engine::eTextureFormat::DEPTH;
    _fbDepthConfig.texFormat            = engine::eTextureFormat::DEPTH;
    _fbDepthConfig.texPixelDataType     = engine::ePixelDataType::UINT_32;
    _fbDepthConfig.texWrapU             = engine::eTextureWrap::REPEAT;
    _fbDepthConfig.texWrapV             = engine::eTextureWrap::REPEAT;

    auto _framebuffer = new engine::CFrameBuffer();
    _framebuffer->addAttachment( _fbColorConfig );
    _framebuffer->addAttachment( _fbDepthConfig );

    engine::float32 _quad_buffData[] = {
     /*|  positions |      colors     |     uvs    |*/
        -0.5f, -0.5f, 1.0f, 1.0f, 1.0f, 0.0f, 0.0f,
         0.5f, -0.5f, 1.0f, 1.0f, 1.0f, 1.0f, 0.0f,
         0.5f,  0.5f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f,
        -0.5f,  0.5f, 1.0f, 1.0f, 1.0f, 0.0f, 1.0f
    };

    engine::uint32 _quad_indices[] = {
        0, 1, 2, 
        0, 2, 3
    };

    engine::CVertexBufferLayout _layout = { { "pos", engine::eElementType::Float2, false },
                                            { "color", engine::eElementType::Float3, false },
                                            { "uv", engine::eElementType::Float2, false } };

    auto _quad_vbuffer = new engine::CVertexBuffer( _layout,
                                                    engine::eBufferUsage::STATIC,
                                                    sizeof( _quad_buffData ),
                                                    _quad_buffData );;

    auto _quad_ibuffer = new engine::CIndexBuffer( engine::eBufferUsage::STATIC,
                                                   6, _quad_indices );

    auto _quad_varray = new engine::CVertexArray();
    _quad_varray->addVertexBuffer( _quad_vbuffer );
    _quad_varray->setIndexBuffer( _quad_ibuffer );

    /* setup a simple scene to render to our target *********************************/

    auto _scene = _app->scene();

    auto _cameraProjData = engine::CCameraProjData();
    _cameraProjData.projection  = engine::eCameraProjection::PERSPECTIVE;
    _cameraProjData.fov         = 45.0f;
    _cameraProjData.aspect      = engine::COpenGLApp::GetWindow()->aspect();
    _cameraProjData.zNear       = 0.1f;
    _cameraProjData.zFar        = 100.0f;

    auto _camera = new engine::COrbitCamera( "orbit",
                                             { 4.0f, 4.0f, 4.0f },
                                             { 0.0f, 0.0f, 0.0f },
                                             engine::eAxis::Z,
                                             _cameraProjData,
                                             engine::COpenGLApp::GetWindow()->width(),
                                             engine::COpenGLApp::GetWindow()->height() );
    _scene->addCamera( _camera );

    auto _light = new engine::LLightDirectional( { 0.8f, 0.8f, 0.8f }, { 0.8f, 0.8f, 0.8f },
                                                 { 0.3f, 0.3f, 0.3f }, 0, { 0.0f, 0.0f, -1.0f } );
    _light->setVirtualPosition( { 5.0f, 0.0f, 5.0f } );
    _scene->addLight( _light );

    auto _plane = engine::CMeshBuilder::createPlane( 10.0f, 10.0f );
    _plane->getMaterial()->setColor( { 0.2f, 0.3f, 0.4f } );
    _scene->addRenderable( _plane );

    auto _box = engine::CMeshBuilder::createBox( 0.25f, 0.5f, 1.0f );
    _box->getMaterial()->setColor( { 0.7f, 0.5f, 0.3f } );
    _box->pos = { 1.0f, 1.0f, 1.0f };
    _scene->addRenderable( _box );

    std::cout << engine::toString( _framebuffer->getConfigAttachment( "color_attachment" ) ) << std::endl;
    std::cout << engine::toString( _framebuffer->getConfigAttachment( "depth_attachment" ) ) << std::endl;

    std::cout << engine::toString( *_framebuffer->getTextureAttachment( "color_attachment" )->data() ) << std::endl;
    std::cout << engine::toString( *_framebuffer->getTextureAttachment( "depth_attachment" )->data() ) << std::endl;

    _app->renderer()->disableShadows();

    while( _app->isActive() )
    {
        engine::CDebugDrawer::DrawLine( { 0.0f, 0.0f, 0.0f }, { 5.0f, 0.0f, 0.0f }, { 1.0f, 0.0f, 0.0f } );
        engine::CDebugDrawer::DrawLine( { 0.0f, 0.0f, 0.0f }, { 0.0f, 5.0f, 0.0f }, { 0.0f, 1.0f, 0.0f } );
        engine::CDebugDrawer::DrawLine( { 0.0f, 0.0f, 0.0f }, { 0.0f, 0.0f, 5.0f }, { 0.0f, 0.0f, 1.0f } );

        if ( engine::CInputHandler::IsKeyDown( ENGINE_KEY_ESCAPE ) )
            break;

        /* Render to a custom target given by our framebuffer */

        // set the render target to our framebuffer
        _framebuffer->bind();

        glEnable( GL_DEPTH_TEST );

        _app->begin();

        // let the renderer to the thing :D
        _app->update();

        // release our render targets
        _framebuffer->unbind();

        /* Render what's inside our framebuffer into a quad on the screen */

        glDisable( GL_DEPTH_TEST );
        glClearColor( 0.5f, 0.5f, 0.5f, 1.0f );
        glClear( GL_COLOR_BUFFER_BIT );
        auto _shaderTex2d = engine::CShaderManager::GetCachedShader( "basic2d_textures" );
        ENGINE_ASSERT( _shaderTex2d, "No shader to render our render target texture into a quad :(" );

        _shaderTex2d->bind();
        _framebuffer->getTextureAttachment( "color_attachment" )->bind();
        _quad_varray->bind();

        glDrawElements( GL_TRIANGLES, _quad_varray->indexBuffer()->count(), GL_UNSIGNED_INT, 0 );

        _quad_varray->unbind();
        _framebuffer->getTextureAttachment( "color_attachment" )->unbind();
        _shaderTex2d->unbind();

        _app->end();
    }

    return 0;
}

File: examples/python/core/sample_framebuffer.py
#!/usr/bin/env python

from tinyrenderer.core import engine
import numpy as np
from OpenGL.GL import *

if __name__ == '__main__' :
    app = engine.Application()

    fbColorConfig = engine.AttachmentConfig()
    fbColorConfig.name                 = 'color_attachment'
    fbColorConfig.attachment           = engine.FboAttachment.COLOR
    fbColorConfig.texInternalFormat    = engine.TextureFormat.RGB
    fbColorConfig.texFormat            = engine.TextureFormat.RGB
    fbColorConfig.texPixelDataType     = engine.PixelDataType.UINT_8
    fbColorConfig.texWrapU             = engine.TextureWrap.REPEAT
    fbColorConfig.texWrapV             = engine.TextureWrap.REPEAT

    fbDepthConfig = engine.AttachmentConfig()
    fbDepthConfig.name                 = 'depth_attachment'
    fbDepthConfig.attachment           = engine.FboAttachment.DEPTH
    fbDepthConfig.texInternalFormat    = engine.TextureFormat.DEPTH
    fbDepthConfig.texFormat            = engine.TextureFormat.DEPTH
    fbDepthConfig.texPixelDataType     = engine.PixelDataType.UINT_32
    fbDepthConfig.texWrapU             = engine.TextureWrap.REPEAT
    fbDepthConfig.texWrapV             = engine.TextureWrap.REPEAT

    framebuffer = engine.FrameBuffer( app.window().width, app.window().height )
    framebuffer.addAttachment( fbColorConfig )
    framebuffer.addAttachment( fbDepthConfig )

    quad_buffData = np.array( [ 
    ##|  positions |    colors    |   uvs   |##
        -0.5, -0.5, 1.0, 1.0, 1.0, 0.0, 0.0,
         0.5, -0.5, 1.0, 1.0, 1.0, 1.0, 0.0,
         0.5,  0.5, 1.0, 1.0, 1.0, 1.0, 1.0,
        -0.5,  0.5, 1.0, 1.0, 1.0, 0.0, 1.0
    ], dtype = np.float32 )

    quad_indices = np.array( [
        0, 1, 2, 
        0, 2, 3
    ], dtype = np.uint32 )

    layout = engine.VertexBufferLayout( [ [ 'pos'  , engine.ElementType.Float2, False ],
                                      [ 'color', engine.ElementType.Float3, False ],
                                      [ 'uv'   , engine.ElementType.Float2, False ] ] )

    quad_vbuffer = engine.VertexBuffer( layout,
                                    engine.BufferUsage.STATIC,
                                    quad_buffData.size * quad_buffData.itemsize,
                                    quad_buffData )

    quad_ibuffer = engine.IndexBuffer( engine.BufferUsage.STATIC, 6, quad_indices )

    quad_varray = engine.VertexArray()
    quad_varray.addVertexBuffer( quad_vbuffer )
    quad_varray.setIndexBuffer( quad_ibuffer )

    #### setup a simple scene to render to our target *********************************

    scene = app.scene()

    cameraProjData = engine.CameraProjData()
    cameraProjData.projection  = engine.CameraProjection.PERSPECTIVE
    cameraProjData.fov         = 45.0
    cameraProjData.aspect      = app.window().aspect
    cameraProjData.zNear       = 0.1
    cameraProjData.zFar        = 100.0

    camera = engine.OrbitCamera( 'orbit',
                             [ 4.0, 4.0, 4.0 ],
                             [ 0.0, 0.0, 0.0 ],
                             engine.Axis.Z,
                             cameraProjData,
                             app.window().width,
                             app.window().height )
    scene.AddCamera( camera )

    light = engine.DirectionalLight( 'directional_1',
                                 [ 0.8, 0.8, 0.8 ],
                                 [ 0.8, 0.8, 0.8 ],
                                 [ 0.3, 0.3, 0.3 ],
                                 [ 0.0, 0.0, -1.0 ] )
    scene.AddLight( light )

    plane = engine.MeshBuilder.CreatePlane( 10.0, 10.0 )
    plane.material.ambient  = [ 0.2, 0.3, 0.4 ]
    plane.material.diffuse  = [ 0.2, 0.3, 0.4 ]
    plane.material.specular = [ 0.2, 0.3, 0.4 ]
    scene.AddRenderable( plane )

    box = engine.MeshBuilder.CreateBox( 0.25, 0.5, 1.0 )
    box.material.ambient  = [ 0.7, 0.5, 0.3 ]
    box.material.diffuse  = [ 0.7, 0.5, 0.3 ]
    box.material.specular = [ 0.7, 0.5, 0.3 ]
    box.position = [ 1.0, 1.0, 1.0 ]
    scene.AddRenderable( box )

    print( 'fbo-config - color_attachment: \n{}'.format( framebuffer.getConfigAttachment( 'color_attachment' ) ) )
    print( 'fbo-config - depth_attachment: \n{}'.format( framebuffer.getConfigAttachment( 'depth_attachment' ) ) )

    print( 'fbo-texture-data - color_attachment: \n{}'.format( framebuffer.getTextureAttachment( 'color_attachment' ).data ) )
    print( 'fbo-texture-data - depth_attachment: \n{}'.format( framebuffer.getTextureAttachment( 'depth_attachment' ).data ) )

    while( app.active() ) :
        engine.DebugDrawer.DrawLine( [ 0.0, 0.0, 0.0 ], [ 5.0, 0.0, 0.0 ], [ 1.0, 0.0, 0.0 ] )
        engine.DebugDrawer.DrawLine( [ 0.0, 0.0, 0.0 ], [ 0.0, 5.0, 0.0 ], [ 0.0, 1.0, 0.0 ] )
        engine.DebugDrawer.DrawLine( [ 0.0, 0.0, 0.0 ], [ 0.0, 0.0, 5.0 ], [ 0.0, 0.0, 1.0 ] )

        if ( engine.InputManager.IsKeyDown( engine.Keys.KEY_ESCAPE ) ) :
            break

        app.update()

        #### Render to a custom target given by our framebuffer ####

        # set the render target to our framebuffer
        framebuffer.bind()

        app.begin()

        # let the renderer to the thing :D
        app.render()

        # release our render targets
        framebuffer.unbind()

        #### Render what's inside our framebuffer into a quad on the screen ####

        glDisable( GL_DEPTH_TEST )
        glClearColor( 0.5, 0.5, 0.5, 1.0 )
        glClear( GL_COLOR_BUFFER_BIT )
        shaderTex2d = engine.ShaderManager.GetCachedShader( 'basic2d_textures' )
        assert shaderTex2d, 'No shader to render our render target texture into a quad :('

        shaderTex2d.bind()
        framebuffer.getTextureAttachment( 'color_attachment' ).bind()
        quad_varray.bind()

        glDrawElements( GL_TRIANGLES, quad_varray.indexBuffer().count(), GL_UNSIGNED_INT, None )

        quad_varray.unbind()
        framebuffer.getTextureAttachment( 'color_attachment' ).unbind()
        shaderTex2d.unbind()
        glEnable( GL_DEPTH_TEST )

        app.end()